### PR TITLE
feat: allow components to opt out of hydration validation

### PR DESCRIPTION
**Pull request by divmain**
## Details

This introduces a mechanism for components to partially or fully opt out from hydration validation. This opt-out only applies to the host element itself, e.g. if a component opts out but there is a mismatch in one of its (non-custom) child elements, hydration will still fail and the host element will be replaced.

A consumer of this mechanism can either specify `static validationOptOut = true;` for full opt-out or `static validationOptOut = ['list', 'of', 'attrs'];` for a partial opt-out (which would permit inequality for the specified attributes only). If, for example, the component author specifies `static validationOptOut = ['foo'];` and the CSR host element's `class` does not match the SSR `class`, hydration will fail and the host element will be replaced.

This PR provides a workaround for the issue described in #3009.

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

* ⚠️ Yes, it does include an observable change.

Technically, if someone had a static `validationOptOut` property on their LWC class, the behavior during SSR rehydration will now be different. However, because SSR is not yet consumed widely, and because this case is so niche, this change should not be a concern.

## GUS work item

W-13044995

### CODE REVIEW

File: packages/@lwc/engine-core/src/framework/hydration.ts [hunk 1]
```diff
@@ -124,6 +133,22 @@ function textNodeContentsAreEqual(node: Node, vnode: VText, renderer: RendererAP
...
 }
 
+function getAttrsToNotValidate(optOutStaticProp: string[] | true | undefined) {
+    if (isUndefined(optOutStaticProp)) {
+        return;
+    }
+    if (isTrue(optOutStaticProp)) {
+        return;
+    }
+    if (!Array.isArray(optOutStaticProp) || !optOutStaticProp.every((el) => isString(el))) {
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/hydration.ts:143]:
```suggestion
    if (!isArray(optOutStaticProp) || !optOutStaticProp.every((el) => isString(el))) {
```

We have an `isArray` in `@lwc/shared`. We should also import an `ArrayEvery` or create it if it doesn't exist.

File: packages/@lwc/engine-core/src/framework/hydration.ts [hunk 2]
```diff
@@ -243,11 +268,25 @@ function hydrateCustomElement(
     vnode: VCustomElement,
     renderer: RendererAPI
 ): Node | null {
-    if (
-        !hasCorrectNodeType<Element>(vnode, elm, EnvNodeTypes.ELEMENT, renderer) ||
-        !isMatchingElement(vnode, elm, renderer)
-    ) {
-        return handleMismatch(elm, vnode, renderer);
+    const { validationOptOut } = vnode.ctor;
+    const attrsToNotValidate = getAttrsToNotValidate(validationOptOut);
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/hydration.ts:272]: I think the rest of this code could be a bit simpler if `getAttrsToNotValidate` returned a function that takes a string as input and outputs a boolean. In the case where `validationOptOut` is falsey, it would just always return true.

There might be a slight perf overhead, but I wouldn't worry about this too much until we have a profiler telling us that it's a problem. Then we can get rid of the `if (!validationOptOut || attrsToNotValidate) {`, the question marks in `attrsToNotValidate?.has?.('class')`, etc.

File: packages/@lwc/engine-core/src/framework/hydration.ts [hunk 3]
```diff
@@ -243,11 +268,25 @@ function hydrateCustomElement(
...
+
+    // The validationOptOut static property can be an array of attribute names.
+    // Any attribute names specified in that array will not be validated, and the
+    // LWC runtime will assume that VDOM attrs and DOM attrs are in sync.
+    //
+    // If validationOptOut is true, no attributes will be checked for correctness
+    // and the runtime will assume VDOM attrs and DOM attrs are in sync.
+    //
+    // Therefore, if validationOptOut is falsey or an array of strings, we need to
+    // examine some or all of the custom element's attributes.
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/hydration.ts:282]: Rather than check for falsiness, I think it would be simpler to just check for `isTrue` and `isArray`. Anything else can be treated as "no filtering expected."

Otherwise, we may need to add tests for every possible falsey value (undefined, null, 0, empty string, etc.), since historically anything we don't test will eventually break.

File: packages/@lwc/engine-core/src/framework/hydration.ts (hunk 1 above)
Comment (divmain) [reply packages/@lwc/engine-core/src/framework/hydration.ts:143]: Addressing this was somewhat complex, due to the type of `Array.prototype.every` once extracted from the prototype. Hopefully the workaround will suffice.

File: packages/@lwc/engine-core/src/framework/hydration.ts (hunk 3 above)
Comment (divmain) [reply packages/@lwc/engine-core/src/framework/hydration.ts:282]: Took care of this with the refactor related to your above comment.

File: packages/@lwc/shared/src/language.ts [hunk 4]
```diff
@@ -75,6 +89,7 @@ export {
     ArraySort,
     ArraySplice,
     ArrayUnshift,
+    arrayEvery,
```
Comment (nolanlawson) [inline packages/@lwc/shared/src/language.ts:92]: Let's use `ArrayEvery` for consistent capitalization

File: packages/@lwc/engine-core/src/framework/hydration.ts [hunk 5]
```diff
@@ -124,6 +135,34 @@ function textNodeContentsAreEqual(node: Node, vnode: VText, renderer: RendererAP
     return false;
 }
 
+// The validationOptOut static property can be an array of attribute names.
+// Any attribute names specified in that array will not be validated, and the
+// LWC runtime will assume that VDOM attrs and DOM attrs are in sync.
+//
+
+//
+
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/hydration.ts:144]:
```suggestion
```

File: packages/@lwc/engine-core/src/framework/hydration.ts [hunk 6]
```diff
@@ -124,6 +135,34 @@ function textNodeContentsAreEqual(node: Node, vnode: VText, renderer: RendererAP
...
+    // If validationOptOut is true, no attributes will be checked for correctness
+    // and the runtime will assume VDOM attrs and DOM attrs are in sync.
+    if (isTrue(optOutStaticProp)) {
+        return (_attrName: string) => false;
+    }
+    // If validationOptOut is an array of strings, attributes specified in the
+    // array will be "opted out". Attributes not specified in the array will still
+    // be validated.
+    if (isArray(optOutStaticProp) && arrayEvery<string>(optOutStaticProp, isString)) {
+        return (attrName: string) => !optOutStaticProp.includes(attrName);
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/hydration.ts:158]: `ArrayIncludes` for consistent usage of prototype methods

File: packages/@lwc/engine-core/src/framework/hydration.ts [hunk 7]
```diff
@@ -371,7 +422,12 @@ function hasCorrectNodeType<T extends Node>(
     return true;
 }
 
-function isMatchingElement(vnode: VBaseElement, elm: Element, renderer: RendererAPI) {
+function isMatchingElement(
+    vnode: VBaseElement,
+    elm: Element,
+    renderer: RendererAPI,
+    shouldValidateAttr: (attrName: string) => boolean = () => true
```
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/hydration.ts:429]: Extracting out `(attrName: string) => boolean` as a `type` may make this look a bit cleaner

**Review by nolanlawson — CHANGES_REQUESTED** [review]
Last few nitpicks, I promise! 😅 LGTM

File: packages/@lwc/shared/src/language.ts (hunk 4 above)
Comment (divmain) [reply packages/@lwc/shared/src/language.ts:92]: I believe that would be more confusing. `arrayEvery !== Array.prototype.every` whereas `ArraySort === Array.prototype.sort`. Exposing `Array.prototype.every` as `ArrayEvery` causes a TypeScript error _in the generated `.d.ts`_, which can't be `@ts-ignore`d.

Since `arrayEvery` is not the same kind of thing as the other `CapitalizedExports`, it seems like capitalizing in that same way would be misleading to future users of `ArrayEvery`.

Comment (nolanlawson) [reply packages/@lwc/shared/src/language.ts:94]: I see, okay. Thanks for the explanation!

**Review by nolanlawson — APPROVED** [review]
